src/logworker.cpp

```diff
@@ -112,14 +114,23 @@ namespace g3 {
 
    void LogWorker::addWrappedSink(std::shared_ptr<g3::internal::SinkWrapper> sink) {
       auto bg_addsink_call = [this, sink] {_impl._sinks.push_back(sink);};
-      auto token_done = g3::spawn_task(bg_addsink_call, _impl._bg.get());
-      token_done.wait();
+      g3::spawn_task(bg_addsink_call, _impl._bg.get());
    }
 
    std::unique_ptr<LogWorker> LogWorker::createLogWorker() {
       return std::unique_ptr<LogWorker>(new LogWorker);
    }
 
+   void LogWorker::removeWrappedSink(std::shared_ptr<g3::internal::SinkWrapper> sink) {
+      auto bg_removesink_call = [this, sink] {
+         _impl._sinks.erase(std::remove(_impl._sinks.begin(), _impl._sinks.end(), sink), _impl._sinks.end());
+
+         // TODO: to be added to the sink API
+         //sink->flush();
+      };
+      g3::spawn_task(bg_removesink_call, _impl._bg.get());
```

Comment:
With the above changes it is less flexible than you suggested but is more robust and easy to reason about.